Mathlib/Topology/Algebra/Valued/ValuativeRel.lean

```diff
@@ -211,10 +238,170 @@ lemma continuous_valuation : Continuous v := by
+automatically an `IsValuativeTopology`. -/
+theorem of_hasBasis {R : Type*} [CommRing R] [ValuativeRel R]
+    {ι : Type*} (p : ι → Prop) (s : ι → Set R)
+    (ih : ∀ [TopologicalSpace R] [IsValuativeTopology R], (nhds 0).HasBasis p s)
+    [τ : TopologicalSpace R] [ContinuousConstVAdd R R] (h : (nhds 0).HasBasis p s) :
+    IsValuativeTopology R := by
+  rw [isValuativeTopology_iff_subgroups_basis_topology_eq]
+  let := (valuation R).subgroups_basis
+  specialize @ih this.topology of_subgroups_basis
+  refine ext_nhds fun x ↦ Filter.ext fun t ↦ ?_
+  rw [← @vadd_mem_nhds_vadd_iff _ _ this.topology _ _ _ _ (-x),
+    ← @vadd_mem_nhds_vadd_iff _ _ τ _ _ _ _ (-x),
+    vadd_eq_add, neg_add_cancel, h.mem_iff, ih.mem_iff]
+
+lemma iff_hasBasis_ne_zero :
+    IsValuativeTopology R ↔
+    (𝓝 (0 : R)).HasBasis (· ≠ 0) fun γ ↦ { x | (valuation R) x < γ } :=
+  iff_hasBasis_zero.trans <| HasBasis.to_hasBasis_iff
+    (fun γ _ ↦ ⟨γ, γ.ne_zero, subset_rfl⟩)
+    (fun _ h ↦ ⟨Units.mk0 _ h, trivial, subset_rfl⟩)
+
+variable (R) in
+lemma hasBasis_nhds_zero_ne_zero [IsValuativeTopology R] :
+    (𝓝 (0 : R)).HasBasis (· ≠ 0) ({ x | (valuation R) x < · }) :=
+  iff_hasBasis_ne_zero.mp inferInstance
+
+alias ⟨_, of_hasBasis_ne_zero⟩ := iff_hasBasis_ne_zero
+
+lemma iff_hasBasis_ne_zero_and_le_one :
+    IsValuativeTopology R ↔
+    (𝓝 (0 : R)).HasBasis (fun γ ↦ γ ≠ 0 ∧ γ ≤ 1) fun γ ↦ { x | (valuation R) x < γ } :=
+  iff_hasBasis_ne_zero.trans <| HasBasis.to_hasBasis_iff
+    (fun γ hγ ↦ ⟨min γ 1, ⟨zero_lt_iff.1 <| lt_min (zero_lt_iff.2 hγ) zero_lt_one,
+      min_le_right _ _⟩, fun x hx ↦ hx.trans_le (a := v x) (min_le_left _ _)⟩)
+    (fun γ hγ ↦ ⟨γ, hγ.1, subset_refl _⟩)
+
+lemma hasBasis_nhds_zero_ne_zero_and_le_one [IsValuativeTopology R] :
+    (𝓝 (0 : R)).HasBasis (fun γ ↦ γ ≠ 0 ∧ γ ≤ 1) ({ x | (valuation R) x < · }) :=
+  iff_hasBasis_ne_zero_and_le_one.mp inferInstance
+
+alias ⟨_, of_hasBasis_ne_zero_and_le_one⟩ := iff_hasBasis_ne_zero_and_le_one
+
+lemma iff_hasBasis_pair :
+    IsValuativeTopology R ↔
+    (𝓝 (0 : R)).HasBasis (fun rs : R × R ↦ 0 <ᵥ rs.1 ∧ 0 <ᵥ rs.2)
+      fun rs ↦ { x | x * rs.2 <ᵥ rs.1 } := by
+  rw [iff_hasBasis_zero]
+  refine HasBasis.to_hasBasis_iff ?_ ?_
+  · simp only [setOf_subset_setOf, Prod.exists, forall_const]
+    intro γ
+    obtain ⟨r, s, h⟩ := exists_valuation_posSubmonoid_div_valuation_posSubmonoid_eq γ
+    refine ⟨r, s, ⟨by simp [← posSubmonoid_def], s.prop⟩, ?_⟩
+    simp only [← h]
+    intro x hx
+    rw [lt_div_iff₀ (by simp [zero_lt_iff])]
+    simp [valuation, hx]
+  · rintro ⟨r, s⟩ ⟨hr, hs⟩
+    refine ⟨Units.mk0 (.mk r ⟨s, hs⟩) ?_, trivial, ?_⟩
+    · simpa using hr
+    · simp [valuation]
+
+lemma hasBasis_nhds_zero_pair [IsValuativeTopology R] :
+    (𝓝 (0 : R)).HasBasis (fun rs : R × R ↦ 0 <ᵥ rs.1 ∧ 0 <ᵥ rs.2)
+      fun rs ↦ { x | x * rs.2 <ᵥ rs.1 } :=
+  iff_hasBasis_pair.mp inferInstance
+
+alias ⟨_, of_hasBasis_pair⟩ := iff_hasBasis_pair
+
+lemma iff_hasBasis_min_inv :
```

Comment:
I would be nice to have a short docstring explaining the name (because it doesn't really follow the naming convention)